Settings: Add new dropdown for choosing analytics behavior

AnuragVasanwala opened the pull request:
## Summary

This PR add support for flexible Tracking Handler. **When SiteKit is installed and active**, in WebStories Settings - it will provide a dropdown with options to select:
- Use SiteKit for analytics _(will be selected as default for compatibility with previous version)_
- Use Web Stories for analytics
- Both

<img width="1379" alt="image" src="https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/302d1873-866a-4e70-9c17-a7e2c11e596d">

## User-facing changes

A dropdown has been added to select the analytics tracking handler. The dropdown is only shown if Site Kit analytics module is active.

Dropdown options:

- Use Site Kit for analytics (default)
- Use only Web Stories for analytics
- Use both

## Testing Instructions

**_Scenario 1_**: Selecting "Use Site Kit for analytics (default)":

- Web Stories will now use Site Kit for tracking analytics.
- Ensure that a note appears indicating that the default option has been selected.
- Verify that the value for the `web_stories_ga_tracking_handler` option changes to `site-kit` upon selecting this option.

**_Scenario 2_**: Selecting "Use only Web Stories for analytics":

- Web Stories will now use inbuilt tracking.
- Ensure that no note appears upon selecting the option.
- Verify that the value for the `web_stories_ga_tracking_handler` option changes to `web-stories`.

**_Scenario 3_**: Selecting "Use both":

- Web Stories will now use both Site Kit and inbuilt tracking.
- Ensure that no note appears upon selecting the option.
- Verify that the value for the `web_stories_ga_tracking_handler` option changes to `both`.

## Reviews

### Does this PR have a security-related impact?

No.

### Does this PR change what data or activity we track or use?

🤔

### Does this PR have a legal-related impact?

No.

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #13506

AnuragVasanwala committed:
✨ Register Tracking Handler Setting

AnuragVasanwala committed:
♻️ Refactor SiteKit `actions` and `filters` on `register` for flexible tracking handler

AnuragVasanwala committed:
✨ Register Tracking Handler API for `wp-dashboard`

AnuragVasanwala committed:
🖍 Register new setting UI with dropdown for flexible Tracking Handler selection

This dropdown will only be visible when SiteKit is installed & active.

AnuragVasanwala committed:
🏗 Add missing doc comment for parameter `$settings`

swissspidy commented on packages/wp-dashboard/src/components/editorSettings/googleAnalytics/index.js:
```diff
@@ -84,6 +85,16 @@ export const TEXT = {
   ),
 };
 
+const ANALYTICS_DROPDOWN_OPTIONS = [
```
The labels need to be translated

swissspidy commented on packages/wp-dashboard/src/components/editorSettings/googleAnalytics/index.js:
```diff
@@ -247,97 +275,110 @@ function GoogleAnalyticsSettings({
+                {__('<b>Note: </b>', 'web-stories')}
+              </TranslateWithMarkup>
+              {TEXT.SITE_KIT_IN_USE}
+            </TextInputHelperText>
```
The `Note:` part should be either part of `TEXT.SITE_KIT_IN_USE` or we should use `sprintf`. No string concatenation.

AnuragVasanwala committed:
♻️ Refactor messages for translation markup

AnuragVasanwala committed:
🧪 Update `editorSettings` unit-test to support for analytics dropdown

AnuragVasanwala committed:
🧪 Update `Site_Kit` unit-test to support for newly added `Setting` parameter to the constructor

AnuragVasanwala committed:
🧪 Add unit-test to `Site_Kit` to support for switching between SiteKit and GA

AnuragVasanwala committed:
Merge branch 'main' into feat/13506-flexible-tracking-support

AnuragVasanwala commented on tests/phpunit/integration/tests/Integrations/Site_Kit.php:
```diff
@@ -54,6 +55,103 @@ public function test_register(): void {
+						return true;
+					}
+					return $blocked;
+				}
```
@swissspidy Is this the correct way to verify static function? This doesn't seem right to me. Would you please suggest possible resolution on this?

swissspidy replied:
It's not possible to compare two separate closures like this. Let's just remove this check here and elsewhere.

The different behavior for the three modes is probably best tested in e2e tests, WDYT?

AnuragVasanwala replied:
> The different behavior for the three modes is probably best tested in e2e tests, WDYT?
Yes, that makes sense.

@swissspidy [Added E2E unit tests](https://github.com/GoogleForCreators/web-stories-wp/pull/13567/commits/82a7b4a4147494487e8b44948e70675cc99c35be) and [removed from PHPUnit](https://github.com/GoogleForCreators/web-stories-wp/pull/13567/commits/860c1b0324ae4ae95e8a6b55b18cad24b82d278e).

Swanand01 replied:
@swissspidy Moved from [E2E Unit Tests](https://github.com/GoogleForCreators/web-stories-wp/pull/13567/commits/82a7b4a4147494487e8b44948e70675cc99c35be) to [`<GoogleAnalyticsSettings />` component Unit Tests](https://github.com/GoogleForCreators/web-stories-wp/pull/13567/commits/383dee834e79fec6a7ed8ea3d7cf1a2ea3c1a95f).

Awaiting for your input on [updating `Note:` message](https://github.com/GoogleForCreators/web-stories-wp/blob/0634e91314aeffb5fa461d84240f0eb54f9efd3f/packages/wp-dashboard/src/components/editorSettings/googleAnalytics/index.js#L81-L84):
https://github.com/GoogleForCreators/web-stories-wp/blob/0634e91314aeffb5fa461d84240f0eb54f9efd3f/packages/wp-dashboard/src/components/editorSettings/googleAnalytics/index.js#L81-L84

cc: @AnuragVasanwala

AnuragVasanwala committed:
Merge branch 'main' into feat/13506-flexible-tracking-support

AnuragVasanwala committed:
⏪ Revert added PHPUnit tests in `Site_Kit`

AnuragVasanwala committed:
🧪 Add E2E unit tests for `Site_Kit`

AnuragVasanwala committed:
Merge branch 'feat/13506-flexible-tracking-support' of https://github.com/GoogleForCreators/web-stories-wp into feat/13506-flexible-tracking-support

AnuragVasanwala committed:
♻️ Display `Site_Kit` note only when it is active

AnuragVasanwala committed:
Merge branch 'main' into feat/13506-flexible-tracking-support

Swanand01 committed:
Remove e2e tests that check if SiteKit message is displayed

Swanand01 committed:
Add component tests for `<GoogleAnalyticsSettings />`:
1. Check if dropdown has rendered
2. Check if SiteKit specific message is displayed
3. Check if `handleUpdateGoogleAnalyticsHandler` is called with the appropriate values

Swanand01 committed:
Export `ANALYTICS_DROPDOWN_OPTIONS` to be used in tests

Swanand01 committed:
Remove unused constants

Swanand01 committed:
Fix eslint issues

Swanand01 committed:
Fix phpcs issues

swissspidy commented on packages/wp-dashboard/src/components/editorSettings/googleAnalytics/index.js:
```diff
@@ -247,97 +291,112 @@ function GoogleAnalyticsSettings({
+                  {TEXT.SITE_KIT_IN_USE}
+                </TranslateWithMarkup>
+              </TextInputHelperText>
+            )}
```
In our last meeting we discussed always displaying this text.

However, after looking at it again, I think we can simply **remove** it. It doesn't add any value in this new layout.

AnuragVasanwala replied:
I think current behaviour seems more clear from user's point of view. Because, note is necessary to convey to the user that `Site Kit` has enabled Google Analytics and will be used to track analytics.

WDYT 🤔 

| Option | Note |
| ------ | ----- |
| Use Site Kit for analytics (default) | <b>Note: </b>Site Kit by Google has already enabled Google Analytics for your Web Stories, all changes to your analytics tracking should occur there. |
| Use only Web Stories for analytics | - |
| Use both | - |

swissspidy replied:
> I think current behaviour seems more clear from user's point of view. Because, note is necessary to convey to the user that `Site Kit` has enabled Google Analytics and will be used to track analytics.

The problem is that this note does not convey this at all. The option in the dropdown does.

That's why the text needs updating and why I don't think it makes sense to show this Note only when the first option is chosen. Either always show it (if Site Kit is active), or never show it.

If we want to rephrase it still, it would make more sense to use something like:

> If Site Kit is active, it will be used to set up Google Analytics by default. However, you can customize the behavior in case you need more flexibility.

AnuragVasanwala replied:
> If Site Kit is active, it will be used to set up Google Analytics by default. However, you can customize the behavior in case you need more flexibility.

Seems to make more sense 👍🏻

Swanand01 replied:
@swissspidy I have updated Site Kit note to be displayed regardless of option chosen in the analytics dropdown.

Commit Ref: https://github.com/GoogleForCreators/web-stories-wp/pull/13567/commits/b86ae3c8e819e7cbc08d056084e53b5dcf8e3ac2

Please let me know if anything else needs to be taken care of. Otherwise, the PR is ready for final review.

AnuragVasanwala committed:
🐛 Fix: Restrict tracking-id input only when `analyticsActive` is active and `analyticsHandler` is `site-kit`

Swanand01 committed:
Fix eslint issues in `googleAnalytics/index.js`

Swanand01 committed:
Add tests to check if the Google Anaytics ID input is enabled/disabled on changing the dropdown option

Swanand01 committed:
Update Site Kit note text and display it regardless of option chosen in dropdown

Swanand01 committed:
Update tests for the Site Kit note

Swanand01 committed:
Remove restriction where measurement ID field is disabled when the "Use Site Kit" option is chosen

Swanand01 committed:
Update tests for measurement ID input

Swanand01 committed:
Merge branch 'main' into feat/13506-flexible-tracking-support

swissspidy commented on packages/wp-dashboard/src/components/editorSettings/googleAnalytics/index.js:
```diff
@@ -34,7 +34,12 @@ import {
+
+/**
+ * External dependencies
+ */
```
This block shouldn't be added.

```suggestion
```

swissspidy commented on packages/wp-dashboard/src/components/editorSettings/googleAnalytics/index.js:
```diff
@@ -247,97 +291,108 @@ function GoogleAnalyticsSettings({
+            placeholder={TEXT.PLACEHOLDER}
+            hasError={Boolean(inputError)}
+            hint={inputError}
+            disabled={analyticsActive}
```
```suggestion
```

Again, we never want to disable the input field, even when Site Kit is active.

Swanand01 replied:
Got it! I thought we didn't want it to be disabled on changing the dropdown. I'll remove the `disabled` property altogether. 
I'll also remove the test that checks if the input is enabled, since it is redundant now that the input is always enabled.

Swanand01 committed:
Remove disable restriction for measurement ID input

Swanand01 committed:
Remove redundant test that checked if the measurement ID input is enabled

Swanand01 committed:
Merge branch 'feat/13506-flexible-tracking-support' of https://github.com/GoogleForCreators/web-stories-wp into feat/13506-flexible-tracking-support
